File: src/test/java/org/railwaystations/rsapi/utils/ImageUtilTest.java
package org.railwaystations.rsapi.utils;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.MethodSource;

import javax.imageio.ImageIO;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.nio.file.Path;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ImageUtilTest {

    @ParameterizedTest
    @CsvSource({ "123123.jpg, jpg",
            "123123.asdfas.jpeg, jpeg",
            "123.Jpg, jpg",
            "123.PNG, png",
            ",",
            "456.docx, docx"})
    void testGetExtension(String filename, String extension) {
        assertThat(extension).isEqualTo(ImageUtil.getExtension(filename));
    }

    @ParameterizedTest
    @CsvSource({ "image/jpeg, jpg",
            "image/png, png"})
    void testMimeToExtension(String contentType, String extension) {
        assertThat(ImageUtil.mimeToExtension(contentType)).isEqualTo(extension);
    }

    private static Stream<String> invalidContentTypes() {
        return Stream.of("", "   ", null, "image/svg", "application/json");
    }

    @ParameterizedTest
    @MethodSource("invalidContentTypes")
    void testMimeToExtensionException(String contentType) {
        assertThat(ImageUtil.mimeToExtension(contentType)).isNull();
    }

    @ParameterizedTest
    @CsvSource({ "jpg, image/jpeg",
            "jpeg, image/jpeg",
            "png, image/png"})
    void testExtensionToMimeType(String extension, String mimeType) {
        assertThat(mimeType).isEqualTo(ImageUtil.extensionToMimeType(extension));
    }

    private static Stream<String> invalidExtensions() {
        return Stream.of("", "   ", null, "image/svg", "application/json");
    }

    @ParameterizedTest
    @MethodSource("invalidExtensions")
    void testExtensionToMimeTypeException(String extension) {
        assertThatThrownBy(() -> ImageUtil.extensionToMimeType(extension)).isInstanceOf(IllegalArgumentException.class);
    }

    @ParameterizedTest
    @CsvSource({ "test.jpg, 100, 100",
            "test.png, 150, 150",
            "test.jpg,, 200",
            "test.png,, 200"})
    void testScaleImage(String filename, Integer newWidth, int expectedWidth) throws IOException {
        var photo = Path.of("src/test/resources", filename);
        var scaledBytes = ImageUtil.scalePhoto(photo, newWidth);
        var scaledImage = ImageIO.read(new ByteArrayInputStream(scaledBytes));
        assertThat(expectedWidth).isEqualTo(scaledImage.getWidth());
        assertThat(expectedWidth).isEqualTo(scaledImage.getHeight());
    }

}

File: core/src/test/kotlin/org/railwaystations/rsapi/core/utils/ImageUtilTest.kt
package org.railwaystations.rsapi.core.utils

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.junit.jupiter.params.provider.MethodSource
import org.railwaystations.rsapi.core.utils.ImageUtil.extensionToMimeType
import org.railwaystations.rsapi.core.utils.ImageUtil.getExtension
import org.railwaystations.rsapi.core.utils.ImageUtil.mimeToExtension
import org.railwaystations.rsapi.core.utils.ImageUtil.scalePhoto
import java.io.ByteArrayInputStream
import java.nio.file.Path
import java.util.stream.Stream
import javax.imageio.ImageIO

internal class ImageUtilTest {
    @ParameterizedTest
    @CsvSource(
        "123123.jpg, jpg", "123123.asdfas.jpeg, jpeg", "123.Jpg, jpg", "123.PNG, png", ",", "456.docx, docx"
    )
    fun testGetExtension(filename: String?, extension: String?) {
        assertThat(extension).isEqualTo(getExtension(filename))
    }

    @ParameterizedTest
    @CsvSource(
        "image/jpeg, jpg", "image/png, png"
    )
    fun testMimeToExtension(contentType: String, extension: String) {
        assertThat(mimeToExtension(contentType)).isEqualTo(extension)
    }

    @ParameterizedTest
    @MethodSource("invalidContentTypes")
    fun testMimeToExtensionException(contentType: String?) {
        assertThat(mimeToExtension(contentType)).isNull()
    }

    @ParameterizedTest
    @CsvSource(
        "jpg, image/jpeg", "jpeg, image/jpeg", "png, image/png"
    )
    fun testExtensionToMimeType(extension: String, mimeType: String) {
        assertThat(mimeType).isEqualTo(extensionToMimeType(extension))
    }

    @ParameterizedTest
    @MethodSource("invalidExtensions")
    fun testExtensionToMimeTypeException(extension: String?) {
        assertThatThrownBy { extensionToMimeType(extension) }
            .isInstanceOf(
                IllegalArgumentException::class.java
            )
    }

    @ParameterizedTest
    @CsvSource(
        "test.jpg, 100, 100", "test.png, 150, 150", "test.jpg,, 200", "test.png,, 200"
    )
    fun testScaleImage(filename: String, newWidth: Int?, expectedWidth: Int) {
        val photo = Path.of("src/test/resources", filename)
        val scaledBytes = scalePhoto(photo, newWidth)
        val scaledImage = ImageIO.read(ByteArrayInputStream(scaledBytes))
        assertThat(expectedWidth).isEqualTo(scaledImage.width)
        assertThat(expectedWidth).isEqualTo(scaledImage.height)
    }

    companion object {

        @JvmStatic
        private fun invalidContentTypes(): Stream<String> {
            return Stream.of("", "   ", null, "image/svg", "application/json")
        }

        @JvmStatic
        private fun invalidExtensions(): Stream<String> {
            return Stream.of("", "   ", null, "image/svg", "application/json")
        }
    }
}
